### PR TITLE
ECO-573 AmazonPay partial processing

### PR DESCRIPTION
**Pull request by profuel**
Changes covering scope of following tickets:
https://spryker.atlassian.net/browse/ECO-638
https://spryker.atlassian.net/browse/ECO-746
https://spryker.atlassian.net/browse/ECO-573
https://spryker.atlassian.net/browse/ECO-562

### CODE REVIEW

File: src/SprykerEco/Shared/Amazonpay/AmazonpayConfig.php [hunk 1]
```diff
@@ -118,4 +118,36 @@ public function getWidgetScriptPathSandbox()
         return $this->get(AmazonpayConstants::WIDGET_SCRIPT_PATH_SANDBOX);
     }
 
+    /**
+     * @return bool
+     */
+    public function getPopupLogin()
+    {
+        return (bool) $this->get(AmazonpayConstants::WIDGET_POPUP_LOGIN);
```
Comment (dereuromark) [inline src/SprykerEco/Shared/Amazonpay/AmazonpayConfig.php]: The CS checker will not like the space after casts.

PS: if this is bool, Core usually recommends `isPopupLogin` or "has..." as method name.

Comment (profuel) [reply src/SprykerEco/Shared/Amazonpay/AmazonpayConfig.php]: Running CS now :)
thanks!

File: src/SprykerEco/Shared/Amazonpay/Transfer/amazonpay.transfer.xml [hunk 2]
```diff
@@ -188,11 +155,12 @@
         <property name="BillingSameAsShipping" type="bool" />
         <property name="OrderReference" type="string" />
         <property name="Email" type="string" /><!-- WHY ??? -->
```
Comment (aurisl) [inline src/SprykerEco/Shared/Amazonpay/Transfer/amazonpay.transfer.xml]: Do you still need this comment?

File: src/SprykerEco/Yves/Amazonpay/Controller/PaymentController.php [hunk 3]
```diff
@@ -7,16 +7,15 @@
 
 namespace SprykerEco\Yves\Amazonpay\Controller;
 
-use InvalidArgumentException;
 use Generated\Shared\Transfer\AmazonpayPaymentTransfer;
 use Generated\Shared\Transfer\QuoteTransfer;
+use InvalidArgumentException;
```
Comment (aurisl) [inline src/SprykerEco/Yves/Amazonpay/Controller/PaymentController.php]: Avoid generic exceptions.

File: src/SprykerEco/Yves/Amazonpay/Controller/PaymentController.php [hunk 4]
```diff
@@ -187,9 +186,12 @@ public function confirmPurchaseAction(Request $request)
         }
 
         if (!$quoteTransfer->getAmazonpayPayment()->getResponseHeader()->getIsSuccess()) {
-            return $this->getFailedRedirectResponse(
-                $quoteTransfer->getAmazonpayPayment()->getResponseHeader()->getErrorCode()
+            $this->addErrorMessage(
+                $quoteTransfer->getAmazonpayPayment()->getResponseHeader()->getErrorMessage()
+                ?? 'amazonpay.payment.failed'
```
Comment (aurisl) [inline src/SprykerEco/Yves/Amazonpay/Controller/PaymentController.php]: Remove string literal, extract to constant.

File: src/SprykerEco/Yves/Amazonpay/Theme/default/layout.twig [hunk 5]
```diff
@@ -23,5 +23,6 @@
 {% endblock %}
 
 {% block javascript %}
+    <script src="https://code.jquery.com/jquery-3.1.1.min.js"></script>
```
Comment (aurisl) [inline src/SprykerEco/Yves/Amazonpay/Theme/default/layout.twig:26]: We have jquery in demoshop. Generally vendor javascript shold be included with package.json

File: src/SprykerEco/Yves/Amazonpay/Theme/default/widget/pay-button.twig [hunk 6]
```diff
@@ -23,16 +23,25 @@
...
             authorization: function() {
-                loginOptions = {scope: 'profile postal_code payments:widget payments:shipping_address payments:billing_address', popup: true};
-                authRequest = amazon.Login.authorize (loginOptions, function(response) {
-                    addressConsentToken = response.access_token;
-                });
+                loginOptions = {
+                    scope: 'profile postal_code payments:widget payments:shipping_address payments:billing_address',
+                    popup: {{ amazonpayConfig.popupLogin ? 'true' : 'false' }}
+                };
+                authRequest = amazon.Login.authorize (loginOptions, {% spaceless %}
```
Comment (aurisl) [inline src/SprykerEco/Yves/Amazonpay/Theme/default/widget/pay-button.twig:36]: Javscript should go to assets folder.

File: src/SprykerEco/Zed/Amazonpay/Business/AmazonpayBusinessFactory.php [hunk 7]
```diff
@@ -171,11 +171,19 @@ protected function getMessengerFacade()
     }
 
     /**
-     * @return PaymentAmazonpayConverterInterface
+     * @return \SprykerEco\Zed\Amazonpay\Business\Converter\AmazonpayConverter
```
Comment (aurisl) [inline src/SprykerEco/Zed/Amazonpay/Business/AmazonpayBusinessFactory.php]: Interface

File: src/SprykerEco/Zed/Amazonpay/Business/Api/Converter/AbstractResponseParserConverter.php [hunk 8]
```diff
@@ -110,14 +117,15 @@ protected function extractHeader(ResponseInterface $responseParser)
...
-            $header->setRequestId($metadata['RequestId']);
+            $header->setRequestId($metadata[self::REQUEST_ID]);
         }
 
-        if (!empty($responseParser->toArray()['Error'])) {
-            $responseArray = $responseParser->toArray();
-            $header->setErrorMessage($responseArray['Error']['Message']);
-            $header->setErrorCode($responseArray['Error']['Code']);
-            $header->setRequestId($responseArray['RequestId']);
+        $responseArray = $responseParser->toArray();
```
Comment (aurisl) [inline src/SprykerEco/Zed/Amazonpay/Business/Api/Converter/AbstractResponseParserConverter.php]: This whole code block can go to method. 
Use `static::` whenever you using class constants.
Why only some of the response data have constants for array keys?

File: src/SprykerEco/Zed/Amazonpay/Business/Api/Converter/AbstractResponseParserConverter.php [hunk 9]
```diff
@@ -128,16 +136,17 @@ protected function extractHeader(ResponseInterface $responseParser)
...
      */
     protected function extractErrorFromConstraints(AmazonpayResponseHeaderTransfer $header)
     {
         if ($header->getConstraints()->count() === 0) {
             return;
         }
 
-        /** @var AmazonpayResponseConstraintTransfer $constraint */
-        $constraint = $header->getConstraints()->offsetGet(0);
+        $constraint = $header->getConstraints()[0];
```
Comment (aurisl) [inline src/SprykerEco/Zed/Amazonpay/Business/Api/Converter/AbstractResponseParserConverter.php]: I would extract to method and give name to it  which would indicate why/what you pick from array first index.

File: src/SprykerEco/Zed/Amazonpay/Business/Api/Converter/AbstractResponseParserConverter.php [hunk 10]
```diff
@@ -128,16 +136,17 @@ protected function extractHeader(ResponseInterface $responseParser)
...
         $header->setErrorCode('amazonpay.payment.error.' . $constraint->getConstraintId());
```
Comment (aurisl) [inline src/SprykerEco/Zed/Amazonpay/Business/Api/Converter/AbstractResponseParserConverter.php]: extract literal

File: src/SprykerEco/Zed/Amazonpay/Business/Api/Converter/AbstractResponseParserConverter.php [hunk 11]
```diff
@@ -178,21 +185,21 @@ protected function extractConstraints(ResponseInterface $responseParser)
...
         $constraintTransfers = new ArrayObject();
 
-        if (empty($result['OrderReferenceDetails']['Constraints'])) {
+        if (empty($result[self::ORDER_REFERENCE_DETAILS][self::CONSTRAINTS])) {
             return $constraintTransfers;
         }
 
-        if (count($result['OrderReferenceDetails']['Constraints']) === 1) {
-            $constraints = array_values($result['OrderReferenceDetails']['Constraints']);
+        if (count($result[self::ORDER_REFERENCE_DETAILS][self::CONSTRAINTS]) === 1) {
```
Comment (aurisl) [inline src/SprykerEco/Zed/Amazonpay/Business/Api/Converter/AbstractResponseParserConverter.php]: Extract to method.

File: src/SprykerEco/Zed/Amazonpay/Business/Api/Converter/AbstractResponseParserConverter.php [hunk 12]
```diff
@@ -178,21 +185,21 @@ protected function extractConstraints(ResponseInterface $responseParser)
...
+            $constraints = array_values($result[self::ORDER_REFERENCE_DETAILS][self::CONSTRAINTS]);
         } else {
-            $constraints = $result['OrderReferenceDetails']['Constraints'];
+            $constraints = $result[self::ORDER_REFERENCE_DETAILS][self::CONSTRAINTS];
         }
 
         foreach ($constraints as $constraint) {
-            if ((!empty($constraint['ConstraintID'])) && !empty($constraint['Description'])) {
+            if ((!empty($constraint[self::CONSTRAINT_ID])) && !empty($constraint[self::DESCRIPTION])) {
                 $constraintTransfer = new AmazonpayResponseConstraintTransfer();
```
Comment (aurisl) [inline src/SprykerEco/Zed/Amazonpay/Business/Api/Converter/AbstractResponseParserConverter.php]: Extract to method.

File: src/SprykerEco/Zed/Amazonpay/Business/Payment/Handler/Transaction/AbstractAmazonpayTransaction.php [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+        $this->apiResponse = $this->executionAdapter->call($amazonpayCallTransfer);
+        $amazonpayCallTransfer->getAmazonpayPayment()
+            ->fromArray($this->apiResponse->modifiedToArray(), true)
+            ->setResponseHeader($this->apiResponse->getHeader());
+
+        $this->transactionsLogger->log(
+            $amazonpayCallTransfer->getAmazonpayPayment()->getOrderReferenceId(),
+            $this->apiResponse->getHeader()
+        );
+        // TODO: remove this hidden call
```
Comment (aurisl) [inline src/SprykerEco/Zed/Amazonpay/Business/Payment/Handler/Transaction/AbstractAmazonpayTransaction.php]: remove todo. fix code.

File: src/SprykerEco/Zed/Amazonpay/Business/Payment/Handler/Transaction/AbstractAmazonpayTransaction.php [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
+     * @param \Generated\Shared\Transfer\AmazonpayCallTransfer $amazonpayCallTransfer
+     *
+     * @throws \Exception
+     *
+     * @return void
+     */
+    protected function loadPaymentEntity(AmazonpayCallTransfer $amazonpayCallTransfer)
+    {
+        if ($this->paymentEntity) {
+            throw new Exception('paymentEntity was previously defined!');
```
Comment (aurisl) [inline src/SprykerEco/Zed/Amazonpay/Business/Payment/Handler/Transaction/AbstractAmazonpayTransaction.php]: Use Custom exceptions.

File: src/SprykerEco/Zed/Amazonpay/Business/Payment/Handler/Transaction/AuthorizeTransaction.php [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+    public function execute(AmazonpayCallTransfer $amazonpayCallTransfer)
+    {
+        $authReferenceId = $this->generateOperationReferenceId($amazonpayCallTransfer);
+        $amazonpayCallTransfer->getAmazonpayPayment()
+            ->getAuthorizationDetails()
+            ->setAuthorizationReferenceId($authReferenceId);
+
+        $amazonpayCallTransfer = parent::execute($amazonpayCallTransfer);
+
+        if ($amazonpayCallTransfer->getAmazonpayPayment()->getResponseHeader()->getIsSuccess()) {
```
Comment (aurisl) [inline src/SprykerEco/Zed/Amazonpay/Business/Payment/Handler/Transaction/AuthorizeTransaction.php]: invert condition

File: src/SprykerEco/Zed/Amazonpay/Business/Payment/Handler/Transaction/AuthorizeTransaction.php [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+    }
+
+    /**
+     * @param \Generated\Shared\Transfer\AmazonpayCallTransfer $amazonpayCallTransfer
+     *
+     * @return string
+     */
+    protected function buildErrorCode(AmazonpayCallTransfer $amazonpayCallTransfer)
+    {
+        return 'amazonpay.payment.error.' .
```
Comment (aurisl) [inline src/SprykerEco/Zed/Amazonpay/Business/Payment/Handler/Transaction/AuthorizeTransaction.php]: extract literal

File: src/SprykerEco/Zed/Amazonpay/Business/Payment/Handler/Transaction/TransactionFactory.php [hunk 17]
```diff
@@ -294,18 +332,19 @@ public function createHandleDeclinedOrderTransaction()
     }
 
     /**
-     * @return \SprykerEco\Zed\Amazonpay\Business\Payment\Handler\Transaction\QuoteTransactionCollection
+     * @return \SprykerEco\Zed\Amazonpay\Business\Payment\Handler\Transaction\TransactionCollection
```
Comment (aurisl) [inline src/SprykerEco/Zed/Amazonpay/Business/Payment/Handler/Transaction/TransactionFactory.php]: interface

File: src/SprykerEco/Zed/Amazonpay/Business/Payment/Handler/Transaction/TransactionFactoryInterface.php [hunk 18]
```diff
@@ -11,17 +11,17 @@
 {
 
     /**
-     * @return \SprykerEco\Zed\Amazonpay\Business\Payment\Handler\Transaction\QuoteTransactionInterface
+     * @return \SprykerEco\Zed\Amazonpay\Business\Payment\Handler\Transaction\AbstractAmazonpayTransaction
```
Comment (aurisl) [inline src/SprykerEco/Zed/Amazonpay/Business/Payment/Handler/Transaction/TransactionFactoryInterface.php:14]: Why remove interface?, same for following.

File: src/SprykerEco/Zed/Amazonpay/Business/Payment/Handler/Transaction/UpdateOrderAuthorizationStatusTransaction.php [hunk 19]
```diff
@@ -8,76 +8,94 @@
...
-            if ($this->apiResponse->getAuthorizationDetails()->getIdList()) {
-                $this->paymentEntity->setAmazonCaptureId(
-                    $this->apiResponse->getAuthorizationDetails()->getIdList()
-                );
+        $amazonPayment = $amazonpayCallTransfer->getAmazonpayPayment();
 
-                $this->paymentEntity->setStatus(
-                    AmazonpayConstants::OMS_STATUS_CAPTURE_COMPLETED
-                );
+        if ($amazonPayment->getResponseHeader()->getIsSuccess()) {
```
Comment (aurisl) [inline src/SprykerEco/Zed/Amazonpay/Business/Payment/Handler/Transaction/UpdateOrderAuthorizationStatusTransaction.php]: Invert conditions

File: src/SprykerEco/Zed/Amazonpay/Communication/AmazonpayCommunicationFactory.php [hunk 20]
```diff
@@ -27,6 +28,14 @@ public function getSalesFacade()
     }
 
     /**
+     * @return \SprykerEco\Zed\Amazonpay\Dependency\Facade\AmazonpayToOmsBridge
```
Comment (aurisl) [inline src/SprykerEco/Zed/Amazonpay/Communication/AmazonpayCommunicationFactory.php]: interface

File: src/SprykerEco/Zed/Amazonpay/Business/Quote/ShipmentDataQuoteInitializer.php [hunk 21]
```diff
@@ -21,9 +20,9 @@ class ShipmentDataQuoteInitializer implements QuoteUpdaterInterface
      */
     public function update(QuoteTransfer $quoteTransfer)
     {
-        $shipmentMethod = new ShipmentMethodTransfer();
+//        $shipmentMethod = new ShipmentMethodTransfer();
```
Comment (aurisl) [inline src/SprykerEco/Zed/Amazonpay/Business/Quote/ShipmentDataQuoteInitializer.php]: Remove commented code.

File: src/SprykerEco/Zed/Amazonpay/Communication/Plugin/Oms/Command/AbstractAmazonpayCommandPlugin.php [hunk 22]
```diff
@@ -7,20 +7,16 @@
 
 namespace SprykerEco\Zed\Amazonpay\Communication\Plugin\Oms\Command;
```
Comment (aurisl) [inline src/SprykerEco/Zed/Amazonpay/Communication/Plugin/Oms/Command/AbstractAmazonpayCommandPlugin.php:8]: Thats too much logic for a command plugins, extract this to facade call.

File: src/SprykerEco/Zed/Amazonpay/Dependency/Client/AmazonpayToCalculationBridge.php [hunk 23]
```diff
@@ -18,6 +18,9 @@ class AmazonpayToCalculationBridge implements AmazonpayToCalculationInterface
      */
     protected $calculationClient;
 
+    /**
+     * @param \Spryker\Client\Calculation\CalculationClientInterface $calculationClient
+     */
     public function __construct(CalculationClientInterface $calculationClient)
```
Comment (aurisl) [inline src/SprykerEco/Zed/Amazonpay/Dependency/Client/AmazonpayToCalculationBridge.php]: Bridges should not have interface in constructor, only type hint as doc block.

**Review by aurisl — CHANGES_REQUESTED** [review]
Generally its bad idea to have busy state machine commands because you can use this logic independently from OMS. Its better have facade method, with logic in Business. Communication should delegate calls mostly.

File: .scrutinizer.yml [hunk 24]
```diff
@@ -1,4 +1,9 @@
 build:
+    dependencies:
+        before:
+            - composer require --dev spryker/console
+            - composer require --dev spryker/transfer
```
Comment (dereuromark) [inline .scrutinizer.yml]: Why not putting those in composer json?

Comment (profuel) [reply .scrutinizer.yml]: Since we need this for every repository, and only for Scrutinizer builds.

Comment (profuel) [reply .scrutinizer.yml]: transfer module - right, may be included in `composer.json`

**Pull request closed without merging**